---
title: "bartools: quickstart guide"
author: "Dane Vassiliadis, Henrietta Holze"
output: 
  rmarkdown::html_document:
    toc: true
    number_sections: true
    toc_depth: 3
    highlight: tango 
date: June 27, 2019
vignette: >
  %\VignetteIndexEntry{bartools-quickstart}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: sentence
---

# Bartools Quickstart Guide

## 0. Install and load the `bartools` package

You can install bartools from
[GitHub](https://github.com/DaneVass/bartools):

```{r eval=FALSE}
# first install Bioconductor dependencies
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("edgeR", "limma", "ComplexHeatmap"))

# then install bartools via GitHub
if (!requireNamespace("devtools", quietly = TRUE)) {
  install.packages("devtools")
}
devtools::install_github("DaneVass/bartools", dependencies = TRUE, force = TRUE)
```

```{r setup}
library(bartools)
library(magrittr)
knitr::opts_chunk$set(dev="png")
```

## 1. Importing DNA barcode count data

Raw barcode count data can be thought of similarly to raw integer-based count data from other count based experiments such as RNA-sequencing.
For these data types the `edgeR` package provides an efficient `DGEList` object structure to store sample counts and associated metadata.
`bartools` makes use of this object structure to store and process DNA barcode counts.

### An example barcoding experiment

For this section we will make use of a hypothetical DNA barcoding dataset based on recent unpublished data from the Dawson lab investigating the response of acute myeloid leukaemia (AML) cells to a novel class of MYST acetyltransferase inhibitor described recently in [MacPherson et al. Nature 2019](https://dx.doi.org/10.1038/s41586-019-1835-6).

AML cells were cultured *in vitro*, barcoded using a lentiviral based barcoding library called [SPLINTR](https://dx.doi.org/10.1038/s41586-021-04206-7), and transplanted into three groups of C57BL/6J mice with daily dosing of MYST inhibitor at low or high dose or a corresponding vehicle control.

Barcode containing cells were harvested from the bone marrow of diseased mice and sequenced in technical replicate.

To follow along with this vignette the raw counts tables and sample metadata are included in the `bartools` package.

```{r}
data(test.dge)
```

### Generating a DGEList object from sample counts and metadata

Counts objects defined above can be specified in a sample metadata sheet as shown below.
This is the easiest way to generate a `DGEList` object containing the count information and metadata of interest for a set of barcode sequencing samples.
An example of this process is shown below.

```{r eval = FALSE}
samplesheet <- read.csv(
  system.file(
    "extdata",
    "test_sampletable.csv",
    package = "bartools",
    mustWork = T
  ),
  header = T,
  stringsAsFactors = F
)
samplesheet
```

To read in a sample sheet in excel format:

```{r eval = FALSE}
samplesheet <-
  readxl::read_excel(
    "samplesheet.xlsx"
  )
samplesheet <- as.data.frame(samplesheet)
rownames(samplesheet) <- samplesheet$sample
samplesheet
```

Whatever format your sample sheet is in, make sure that the rownames are your unique sample names. 

Load in the counts as specified in the samplesheet into a DGEList object.
The function expects file locations to be specified either as character vector of filenames, or as a column named files in the samplesheet.

```{r eval = FALSE}
dge <- edgeR::readDGE(
  files = samplesheet,
  group = samplesheet$treatment,
  labels = samplesheet$sample,
  header = F
)
```

This results in the creation of a DGEList object containing counts and metadata information for each sample.

```{r}
# Load the test dataset 
data(test.dge)
test.dge
```

## 2. Data QC and Normalisation

### Data QC

We first want to ensure that we are working with clean data.
Using the `thresholdCounts()` function we can determine an appropriate threshold to apply to the data to maximise signal to noise and retain true and informative barcodes.

We can test different thresholding parameters, such as absolute thresholds on total read counts as below.

```{r}
# Remove barcodes (rows) with no data
test.dge <- test.dge[rowSums(test.dge$counts) != 0, ]
```


```{r}
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 1,
  minSamples = 1,
  plot = T,
  group = "Treatment"
)
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 10,
  minSamples = 1,
  plot = T,
  group = "Treatment"
)
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 10,
  minSamples = 3,
  plot = T,
  group = "Treatment"
)
```

Or relative thresholds based on proportion within a sample.

```{r}
thresholdCounts(
  test.dge,
  type = "relative",
  threshold = 1e-10,
  minSamples = 1,
  plot = T,
  group = "Treatment"
)
thresholdCounts(
  test.dge,
  type = "relative",
  threshold = 1e-5,
  minSamples = 1,
  plot = T,
  group = "Treatment"
)
thresholdCounts(
  test.dge,
  type = "relative",
  threshold = 1e-5,
  minSamples = 3,
  plot = T,
  group = "Treatment"
)
```

Here we will continue with an absolute threshold of 10.

```{r}
dge.filtered <-
  thresholdCounts(
    test.dge,
    type = "absolute",
    threshold = 10,
    minSamples = 2,
    plot = F
  )
```

We then normalise samples to sequencing depth to counts per million using `normaliseCounts()`.

```{r}
dge.cpmnorm <- normaliseCounts(dge.filtered, method = "CPM")
```


We can plot the raw and normalised sequencing depth to get an idea of depth discrepancies between PCR replicates.

```{r}
# raw counts per sample
plotReadCounts(dge.filtered, group = "Treatment")
```

```{r}
# normalised counts per sample
plotReadCounts(dge.cpmnorm, group = "Treatment")
```

For lentiviral based cellular barcoding experiments, such as this one, it is common for the library to exhibit a degree of skewness based on the cloning method.
This means that some barcodes are represented in the library more than others and so have a greater chance to be transduced into multiple cells.\

Most experiments assume that each individual barcode is transduced into only one cell, and that each cell is only transduced with one barcode.
This is ensured using a low multiplicity of infection (MOI) transduction in which the likelihood that a cell is transduced with one or more barcode containing virions follows a Poisson distribution.

With this in mind, it also can be useful to check the total counts per barcode to identify bias in counts in sample vs. frequency of barcode in reference library.\
The barcodes are labelled based on their ranked frequency in the reference library.

```{r}
# plot detected barcodes ordered by frequency in reference library
plotBarcodeCounts(dge.cpmnorm, log10 = F)

# plot log10 barcode counts
plotBarcodeCounts(dge.cpmnorm, log10 = T)

# order barcodes by count across samples
plotBarcodeCounts(dge.cpmnorm, log10 = F, order = T)

# order barcodes by count across samples with log norm
plotBarcodeCounts(dge.cpmnorm, log10 = T, order = T)
```

In the first and second plot individual barcodes on the x-axis are ordered based on their frequency in the reference library pool.\
An increased number of counts per barcode toward the left hand side of the plot would be suggestive of transduction bias, meaning that there are more reads on average attributed to the more abundant barcodes in the library.
And so, likely multiple cells were transduced with the same barcode.\
We don't see this here suggesting that this is not a problem for this experiment.

### Check correlation between PCR replicates

It is also important to ensure that individual samples are sequenced to an appropriate depth as this ensures that the entire barcode repertoire present in a sample is captured in the data.
Sequencing technical duplicates of a sample generated at the library PCR stage is a good way to ensure this.

In our experiment we have 9 samples total, each with two PCR technical replicates.
Here, we correlate the barcode distributions for each pair of technical replicates.  
NB: This only works for paired replicates, i.e. not more than 2. 

```{r message=FALSE}
# get all unique samples
# column "group" contains information on replicates here
unique_samples <- unique(dge.filtered$samples$group)

# only plot subset of samples. 
lapply(unique_samples[1], function(x) {
  # subset dge object to get replicates of sample
  replicate_names <- colnames(dge.filtered)[dge.filtered$samples$group %in% as.character(x)]
  plotBarcodeRegression(
    dge.filtered,
    sample1 = replicate_names[1],
    sample2 = replicate_names[2],
    rug = T,
    trans = "log1p"
  )
})
```

We fit a linear model to both technical replicates per sample and plot the regression line.
Note that we expect a very high correlation because these are PCR replicates of the same barcode pool.

We can also easily get the correlation values between replicates using `calcReplicateCorr`.

Samples can be filtered for high or low correlation using the `threshold` and `return` variables.

```{r}
corrs <- calcReplicateCorr(dge.filtered, group = "Group")
corrs[which(corrs < 0.999)]
corrs
```

Finally sample replicates can be correlated globally using `plotBarcodeCorrelation`.

```{r}
# Pearson correlation, full
plotBarcodeCorrelation(dge.filtered, clustered = T, upper = F, method = "pearson")
plot

# Pearson correlation, upper
plotBarcodeCorrelation(dge.filtered, clustered = T, upper = T, method = "pearson")
```

```{r}
# Spearman correlation
plotBarcodeCorrelation(dge.filtered, clustered = T, upper = T, method = "spearman")
```

### Collapse PCR replicates in object

Now that we know our samples are of good quality we have no further use of the PCR replicate information.
From this point onward its a good idea to collapse our PCR replicates.

```{r}
dim(dge.filtered)
```

`collapseReplicates` can take the average (default behavior) or the sum of PCR technical replicates within each sample.
Here we take the average.
Users may want to sum PCR replicates if there is evidence of sampling bias across technical repeats (i.e. poor correlation score or other evidence).

```{r}
dge.filtered.collapsed <- collapseReplicates(
  dge.filtered,
  group = "group",
  method = "mean"
)
```

The result is a clean barcode sequencing dataset ready for further investigation and visualisation.

```{r}
head(dge.filtered.collapsed)
```

If we repeat the above correlation plots we can see general similarities across samples

```{r}
# Pearson correlation, full
plotBarcodeCorrelation(dge.filtered.collapsed, clustered = T, upper = F, method = "pearson")
```

## 3. Visualisation

`bartools` includes a range of visualisation options for examining barcode-seq datasets.

### Bubble plot

Sometimes a visual depiction of the data is most suitable.
Here, barcodes/tags are represented by bubbles aligned on a single plane.
The size of the bubbles reflects the percentage abundance of each barcode within a sample.

```{r}
plotBarcodeBubble(dge.filtered.collapsed, 
                  proportionCutoff = 10, 
                  labelBarcodes = T)
```

Using the `orderSample` parameter, bubbleplots can also be arranged according to frequency in a particular sample which can help with visual comparison of large vs small clones across samples and conditions.

```{r}
plotOrderedBubble(dge.filtered.collapsed, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  orderSample = "T0", 
                  colorDominant = F, 
                  filterCutoff = 0.001,
                  group = "Treatment")
```

Barcodes that fail to meet a defined abundance threshold in any sample can be greyed out.

```{r}
plotOrderedBubble(dge.filtered.collapsed, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  orderSample = "T0", 
                  colorDominant = T, 
                  filterCutoff = 0.001, 
                  group = "Treatment")
```

Or filtered from the plot entirely using `filterCutoff` parameters

```{r}
plotOrderedBubble(dge.filtered.collapsed, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  orderSample = "T0", 
                  colorDominant = T, 
                  filterCutoff = 0.1,
                  group = "Treatment")
```

### Barcode Plot

Alternatively, we can focus in on the most abundant barcodes within a set of samples to more easily observe how these change in frequency over the course of an experiment.

```{r}
plotBarcodeHistogram(dge.filtered.collapsed, topN = 10, alphaLowFreq = 0)
```

### Timeseries Plots

For timecourse experiments it is useful to visualise how the kinetics of barcode diversity changes over time.
In this instance we can use `plotBarcodeTimeseries` to get an idea of the relative abundance of the top *n* barcodes in a sample relative to others.

```{r}
plotBarcodeTimeseries(dge.filtered.collapsed, 
                      top = 50, 
                      seed = 10101)
```

### Principal Components Analysis

A global level PCA analysis is a good way to get a high level understanding of the similarities and differences between samples.

```{r}
plotBarcodePCA(dge.filtered.collapsed, 
               groups = "Treatment", 
               ntop = 1000,
               pcs=c(1, 2))
```

### Heatmaps

Another method of comparing abundance across samples is using a heatmap.
Here barcodes ranked among the top n most abundant within each sample are indicated by an asterisk.
This heatmap shows high dose samples are generally distinct from the low dose and vehicle group.

```{r}
plotBarcodeHeatmap(
  normaliseCounts(dge.filtered.collapsed, method = "CPM"), 
  topN = 5,
  showBarcodes = F,
  group = "Treatment"
)
```

## 4. Analysing Composition and Diversity

Its important to not only be able to visualise the data but also understand relationships between barcodes/tags at the data level.

### Identify compositional bias within samples

The above plots give a global visualisation of the abundance of each barcode within a sample however the compositional makeup can be obscured by visualising the data in this way.
it can be helpful to examine the fraction of barcodes that comprise a sample.
These plots calculate the cumulative sum of a sample in relation to other samples defined by the user.

```{r}
samples <-
  which(dge.filtered.collapsed$samples$Treatment == "Vehicle")

plotBarcodeCumSum(
  dge.filtered.collapsed,
  referenceSample = "T0",
  samples = colnames(dge.filtered.collapsed$counts)[samples]
)
```

### Identifying abundant barcodes within samples

It is important to be able to determine which barcodes are most abundant within each sample.
`bartools` allows this to be easily calculated according to an abundance threshold.

```{r}
top.bc <- getDominantBarcodes(dge.filtered.collapsed, threshold = 0.05)
top.bc[1:5]
```

We can then use specific plots to visualise the dominance of specific barcodes within and across samples.

```{r}
# plot top barcodes across samples for mouse 10 (High Dose group)
plotBarcodeBoxplot(
  dge.filtered.collapsed,
  barcodes = top.bc$`10_High_dose`,
  group = "Treatment",
  conditions = c("Low_dose", "High_dose", "Vehicle"),
  point = T
)
```

### Calculating and plotting percentile abundance.

The above graphs demonstrate that relatively few barcodes can sometimes comprise the majority of a sample's clonality, particularly following a selective event such as drug treatment.
It is useful to formally analyse this based on a desired percentile threshold.
A common threshold is the 95th percentile.
This can eliminate small barcodes that comprise the tail of the dataset and give a sense of how many clones truly comprise each sample

```{r}
top_barcodes <- calcPercentileBarcodes(dge.filtered.collapsed, percentile = 0.95)
top_barcodes$NumBarcodes
top_barcodes$TopBarcodeCounts$`6_High_dose`
top_barcodes$TopBarcodes$`6_High_dose`
```

We can compare the number of detected barcodes in the top 95th percentile per sample and the total sample.

```{r}
plotDetectedBarcodes(
  dge.filtered.collapsed,
  percentile = 1,
  plot = T,
  group = "Treatment", 
)
plotDetectedBarcodes(
  dge.filtered.collapsed,
  percentile = 0.95,
  plot = T,
  group = "Treatment"
)

plotDetectedBarcodes(
  dge.filtered.collapsed,
  percentile = 0.80,
  plot = T,
  group = "Treatment"
)
```

These plots show that there are few clones that comprise the majority of the dataset per mouse.
Also, there are generally fewer clones present in the high dose group compared to the vehicle or low dose groups.

### Diversity analysis

We can examine within-sample diversity in a few different ways.
The most common are Shannon, Simpson, Inverse Simpson and Gini.
Each will be applicable in different circumstances, however the Shannon diversity index is more widely used to compare global diversity amongst populations of barcoded cells.

`calcDivIndexes` can be used to determine various diversity indices per sample

```{r}
diversity <- calcDivIndexes(dge.filtered.collapsed, group = "Treatment")
diversity
```

These diversity calculations can then be fed to `plotDivIndexes` for visualisation as either a bar or dotplot

```{r}
# bar plot
plotDivIndexes(
  dge.filtered.collapsed,
  div = diversity,
  metric = "shannon",
  group = "Treatment"
)
# dot plot
plotDivIndexes(
  dge.filtered.collapsed,
  div = diversity,
  metric = "shannon",
  group = "Treatment",
  type = "point"
)
# summarized by condition as box plot
plotDivIndexes(
  dge.filtered.collapsed,
  div = diversity,
  metric = "shannon",
  group = "Treatment",
  type = "box"
)
```

### Comparing abundance

We can also statistically test for barcodes / tags that are over / underrepresented in a group of samples relative to another using the internal edgeR framework.
`bartools` contains a convenience wrapper for this functionality

```{r}
compareAbundance(dge.filtered.collapsed,
                 group = "Treatment", 
                 condition1 = "Low_dose",
                 condition2 = "High_dose",
                 pval = 1e-04,
                 logFC = 10)
```

## 5. Session Info

```{r}
sessionInfo()
```
