---
title: "bartools: import and QC single-cell barcode data"
author: "Henrietta Holze"
output: 
  rmarkdown::html_document:
    toc: true
    number_sections: true
    toc_depth: 3
    highlight: tango 
date: September 05, 2023
vignette: >
  %\VignetteIndexEntry{import-qc-single-cell}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: sentence
---

# TODO
create example dataset: counts.tsv, cell-barcode-anno.tsv, seurat object
add description of example dataset. 


# Import and QC of single-cell barcode data

This vignette shows how to
1. Import the barcode counts table produced by BARtab single-cell workflow
2. Creat QC plots and filter barcodes
3. Add barcode counts to a Seurat object
4. Analyze and visualize barcoded single-cell RNA-seq data

If filtering has already been performed by BARtab, using `umi_count_filter` and `umi_fraction_filter` parameters, skip to section 2.  
If the QC thresholds need to be revised, follow section 1. 

## 0. Load the `bartools` package

```{r setup}
library(bartools)
knitr::opts_chunk$set(dev="png")
```

## 1. Import DNA barcode count data

Raw barcode count data can be thought of similarly to raw integer-based count data from other count based experiments such as RNA-sequencing.
For these data types the `edgeR` package provides an efficient `DGEList` object structure to store sample counts and associated metadata.
`bartools` makes use of this object structure to store and process DNA barcode counts.

### An example barcoding experiment

For this section we will make use of a hypothetical DNA barcoding dataset based on recent unpublished data from the Dawson lab investigating the response of acute myeloid leukaemia (AML) cells to a novel class of MYST acetyltransferase inhibitor described recently in [MacPherson et al. Nature 2019](https://dx.doi.org/10.1038/s41586-019-1835-6).

AML cells were cultured *in vitro*, barcoded using a lentiviral based barcoding library called [SPLINTR](https://dx.doi.org/10.1038/s41586-021-04206-7), and transplanted into three groups of C57BL/6J mice with daily dosing of MYST inhibitor at low or high dose or a corresponding vehicle control.

Barcode containing cells were harvested from the bone marrow of diseased mice and sequenced in technical replicate.

To follow along with this vignette the raw counts tables and sample metadata are included in the `bartools` package.

## 2. QC and filter barcodes

First, we load the results of BARtab. Below is an example of how we can load this data using the `readBartabCounts` function. 

```{r, eval = FALSE}
counts_path <- "full/path/to/BARtab/umi.counts.tsv"
counts <- readBartabCounts(counts_path)
```

For now, users can follow along with this vignette using the test dataset included in `bartools`

```{r}
data(test.bartab.sc)
```

This unfiltered barcode counts table is in long format and potentially has multiple rows per cell ID if multiple barcodes per cell are detected. 

We assess the quality of the data by looking at the number of barcodes detected per cell and the number of UMIs supporting the most frequent barcode per cell. 

```{r}
plotBarcodesPerCell(test.bartab.sc)
```

```{r}
plotUmiPerBarcode(test.bartab.sc)
```

In order to identify a suitable threshold for the minimum number of UMIs supporting a barcode, we can plot the cummulative sum of how many cells would pass each threshold. 

```{r}
plotUmiFilterThresholds(test.bartab.sc)
```

If we would remove barcodes that are supported by a single UMI, we would lose half the cells. Therefore, we do not filter based on minimum number of UMIs.

However, we can remove minor barcodes from cells, i.e. barcodes that have less than half the number of supporting UMIs than the major barcode in the cell. 

```{r}
counts_filtered <- filterBarcodes(test.bartab.sc, umiCountFilter = 2, umiFractionFilter = 0.5)
```

The following plot shows that the number of cells with multiple barcodes detected is now reduced. 

If we would want to keep only the most frequent barcode per cell, we could do so with `umi_fraction_filter = 1`. 
Ties will be kept. 


```{r}
plotBarcodesPerCell(counts_filtered)
```

Finally, we aggregate barcodes per cell by concatenating barcodes and UMIs with `;`.  

```{r}
counts_agg <- aggregateBarcodes(counts_filtered)
head(counts_agg)
```

This allows to add the barcode data into the metadata of a `Seurat` or `SingleCellExperiment` object. 


## 2. Add barcode data to Seurat or SingleCellExperiment object

### Barcode metadata import into Seurat

Generate Seurat object from Cell Ranger scRNA expression matrix.
```{r, eval = FALSE}
library(Seurat)
expression.data <-
  Read10X(data.dir = "../../bartools_sc_example/GEX_filtered_feature_bc_matrix/", strip.suffix = T)
sc <- CreateSeuratObject(
  counts = expression.data,
  min.cells = 3,
  min.features = 200,
  project = "bartools_example"
)
sc
```

Add SPLINTR barcode annotation to metadata.
Make sure the cell IDs match. Potentially suffix needs to be added to the BARtab results (e.g. `-1`). 
Add barcode and UMI count columns to the metadata of the SCE object.

```{r, eval = FALSE}
rownames(counts_agg) <- counts_agg$cellid
sc <- Seurat::AddMetaData(sc, counts_agg)
```

### Barcode metadata import into SingleCellExperiment

```{r}
data(test.sce)
colData(test.sce)[c("barcode", "bc.umi.count")] <-
  counts_agg[colnames(test.sce), c("barcode", "bc.umi.count")]
head(colData(test.sce))
```

Group barcode metadata into detected and not.detected
```{r}
test.sce$detected <- ifelse(is.na(test.sce$barcode), "not.detected", "detected")
table(test.sce$detected)
```

## 4. Analyze and visualize barcoded scRNA-seq data

```{r}
data(test.sce)
```

```{r}
message("Percentage of cells with no barcode detected")
length(which(is.na(test.sce$barcode))) / ncol(test.sce) * 100

message("Percentage of cells with a barcode detected")
length(which(!is.na(test.sce$barcode))) / ncol(test.sce) * 100
```

We can repeat the QC plot to check how many barcodes were detected in a cell, including no barcodes detected.

```{r}
plotBarcodesPerCell(test.sce, aggregated = T, sep = ",")
```

We can visualize the distribution of number of cells per clone.

For that, plot all clones, ordered by their size. A clone is here defined as barcode or unique combination of barcodes. 

Usually, more than one barcode detected in a cell could mean a doublet. The number of cells for those clones tend to be low. Seeing more cells per clone with more than one barcode suggests it is truly from two integration events.


```{r}
plotCellsPerGroup(
  test.sce,
  group = "barcode",
  order = T,
  threshold = 5,
  plot = T,
  label = T,
  sep = ","
)
```

To check whether a clone is enriched in a cluster or cell type, we can perform a hypergeometric test.

```{r}
plotClusterEnrichment(
  test.sce,
  group = "barcode",
  factor = "BC_1614",
  clusters = "seurat_clusters",
  threshold = 0.01,
  order = T,
  plot = T
)

```


```{r}
plotCellsInClusters(test.sce, 
                    group = "barcode", 
                    factor = "BC_12904", 
                    clusters = "seurat_clusters")
```


## 5. Session Info

```{r}
sessionInfo()
```
