---
title: "bartools: quickstart guide"
author: "Dane Vassiliadis"
output: 
  rmarkdown::html_document:
    toc: true
    number_sections: true
    toc_depth: 3
    highlight: tango 
date: June 27, 2019
vignette: >
  %\VignetteIndexEntry{bartools-quickstart}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Quickstart Guide

Load the bartools library 
```{r setup}
library(bartools)
```

## A simple analysis
Here we will follow the analysis of an example experiment to understand the basic functions of bartools. Say we have a population of cells growing in culture. These cells were transduced with a lentiviral vector encoding an mCHERRY fluorescent protein and a downstream barcode in the 3' UTR possessing the following structure:

```
tgaccatgtacgattgactaNNSWSNNWSWNNSWSNNWSWNNSWSNNWSWNNSWSNNWSWNNSWSNNWSWNNSWSNNWSWtgctaatgcgtactg
```

Two days after transduction, two sets of 5000 mCHERRY positive cells were sorted using flow cytometry into separate wells and resuspended in culture media. Genomic DNA from the cells was harvested 1 week following transduction and barcode regions were specifically amplified using PCR and sequenced for 75 cycles on an Illumina NextSeq 550 resulting in single ended fastq files containing barcoded reads. 

## Importing DNA barcode count data 
Raw barcode count data can be thought of similarly to raw integer-based count data from other count based experiments such as RNA-sequencing. For these data types the `edgeR` package provides an efficient `DGEList` object structure to store sample counts and associated metadata. `bartools` makes use of this object structure to store and process DNA barcode counts. 

### A real world example barcoding experiment
For this section we will make use of a DNA barcoding dataset generated in a recent study (Bell and Fennell et al. Nat Comms, 2019) that investigated the clonal diversity within populations of acute myeloid leukaemia (AML) cells cultured in the presence or absence of the BET bromodomain inhibitor IBET-151. 

AML cells were cultured in vitro, barcoded using a lentiviral based barcoding library `reference SPLINTR paper here`, and treated with weekly increasing doses of IBET-151 or a corresponding vehicle control. Barcode samples were sequenced for two biological replicates of this experiment at the baseline timepoint, and at week 1 and week 4, corresponding to the IC40 and IC90 of the drug respectively. 

To follow along with this vignette the raw counts tables and sample metadata can be found at `insert link here to the raw datasets`


### Generating a DGEList object from sample counts and metadata
Counts objects defined above can be specified in a sample metadata sheet as shown below. This is the easiest way to generate a `DGEList` object containing the count information and metadata of interest for a set of barcode sequencing samples. 

```{r}
samplesheet <- read.csv(system.file("extdata", "test_sampletable.csv", package = "bartools", mustWork = T), header = T, stringsAsFactors = F)
samplesheet
```

Load in the counts as specified in the samplesheet into a DGEList object
```{r eval=FALSE}
dge <- edgeR::readDGE(files = samplesheet, group = samplesheet$treatment, labels = samplesheet$sample, header = T)
```

This results in the creation of a DGEList object containing counts and metadata information for each sample
```{r}
data(test.dge)
test.dge
```

## Data QC and Normalisation

### Data QC
We first want to ensure that we are working with clean data. 
We remove barcodes that have no counts across all samples. 
```{r}
# Remove rows with no data
keeprows = rowSums(test.dge$counts) >= 1
dge.filtered <- test.dge[keeprows,]
dge.filtered.counts <- dge.filtered$counts
dim(dge.filtered.counts)
```

We then normalise samples to sequencing depth
```{r}
# Normalise each sample to counts per million
dge.cpmnorm <- as.data.frame(cpm(dge.filtered.counts, normalized.lib.sizes = T))
```

We can plot the raw and normalised sequencing depth to get an idea of depth discrepancies between PCR replicates

```{r}
# raw counts per sample
plotReadCounts(dge.filtered.counts, group = dge.filtered$samples$Treatment)
```

```{r}
# normalised counts per sample
plotReadCounts(dge.cpmnorm, group = dge.filtered$samples$Treatment)
```

For lentiviral based barcoding experiments, such as this one, it is common for the library to exhibit a degree of skewness based on the cloning method. 
This means that some barcodes are represented in the library more than others and so have a greater chance to be transduced into multiple cells.
Most experiments assume that each individual barcode is transduced into only one cell, and that each cell is only transduced with one barcode.This is ensured using a low multiplicity of infection (MOI) transduction in which the likelihood that a cell is transduced with one or more barcode containing virions follows Poisson statistics.
With this in mind, it also can be useful to check the total counts per barcode to identify bias in counts in sample vs. frequency of barcode in reference library.

```{r}
# plot detected barcodes
plotBarcodeCounts(dge.cpmnorm, log10 = F)
plotBarcodeCounts(dge.cpmnorm, log10 = T)

```

In this plot individual barcodes on the x-axis are ordered based on their frequency in the library pool. 
An increased number of counts per barcode toward the left hand side of the plot would be suggestive of transduction bias, meaning that there are more reads on average attributed to the more abundant barcodes in the library. And so, likely multiple cells were transduced with the same barcode. We don't see this here suggesting that this is not a problem for this experiment. 

### Check correlation between PCR replicates
It is also important to ensure that individual samples are sequenced to an appropriate depth as this ensures that the entire barcode repertoire present in a sample is captured in the data. Sequencing technical duplicates of a sample generated at the library PCR stage is a good way to ensure this. 

In our experiment we have 9 samples total, each with two PCR technical replicates. 
Here we correlate the barcode distributions for each pair of technical replicates. 

```{r}
samps <- unique(test.dge$samples$Group)
lapply(samps[1:3], function(x){
  df <- test.dge[,test.dge$samples$Group %in% as.character(x)]
  plotBarcodeRegression(df, samp1 = colnames(df)[[1]], samp2 = colnames(df)[[2]])
})
```
We fit a linear model to both technical replicates per sample and plot the regression line. 
Note that we expect a very high correlation because these are PCR duplicates of the same barcode pool. We can easily get the correlation values between replicates
```{r}
corrs <- calcReplicateCorr(test.dge, group = test.dge$samples$Group, corr.thresh = 0.9)
corrs
```

```{r}
which(corrs < 0.9)
```

As we can see there is only a single sample that does not have a R^2 correlation value over our arbitrary threshold of 0.9 (just barely misses the cutoff). We can continue since these samples look pretty good. If there were any samples with low correlation values we might consider removing them from the analysis or analysing them separately. 
  
### Collapse PCR replicates in object
Now that we know which samples are of good quality we no longer need the PCR replicate information. From this point onward its a good idea to collapse our PCR replicates. 
```{r}
dim(dge.filtered)
```

```{r}
# take the average of PCR technical duplicates within a sample
dge.filtered.collapsed <- collapseReplicates(dge.filtered, groupby = dge.filtered$samples$Group, by = "mean")
```

The result is a clean barcode sequencing dataset ready for further investigation and visualisation.
```{r}
dim(dge.filtered.collapsed)
```


## Visualisation

`bartools` includes a range of visualisation options for examining barcode-seq datasets. 

### Bubble plot
Sometimes a visual depiction of the data is most suitable. Here barcodes/tags are represented by bubbles aligned on a single plane. The size of the bubbles reflects the percentage abundance of each barcode within a sample
```{r}
plotBarcodeBubble(dge.filtered.collapsed$counts, proportion.cutoff = 10, labels = T)
```

Bubbleplots can also be ordered according to a particular sample which can help with visual representation of large vs small clones. 
```{r}
plotOrderedBubble(dge.filtered.collapsed$counts, proportion.cutoff = 10, labels = T, orderSample = "T0", colorDominant = T)
plotOrderedBubble(dge.filtered.collapsed$counts, proportion.cutoff = 10, labels = T, orderSample = "T0", colorDominant = F)
```

### Barcode Plot
Alternatively, we can focus in on the most abundant barcodes within a set of samples to more easily observe how these change in frequency over the course of an experiment.  
```{r}
plotBarcodeHistogram(dge.filtered.collapsed$counts, sample = dge.filtered.collapsed$samples$Group[[5]], top = 50)

```

### Timeseries Plots  
For timecourse experiments it is useful to visualise the kinetics of barcode diversity changes over time
```{r}
plotBarcodeTimeseries(dge.filtered.collapsed$counts[,c(1:10)])

```

### Principal Components Analysis
A global level PCA analysis is a good way to get a high level understanding of the similarities and differences between samples. 
```{r}
plotBarcodePCA(dge.filtered.collapsed, intgroup = "Treatment")
plotBarcodePCA(dge.filtered.collapsed[,dge.filtered.collapsed$samples$Treatment %in% c("T0","Vehicle","High_dose")], 
               intgroup = "Treatment")
```

### Heatmaps
```{r}
plotBarcodeHeatmap(dge.filtered.collapsed$counts[,dge.filtered.collapsed$samples$Treatment %in% c("T0", "Vehicle","High_dose")], 
N = 5)
```

## Analysing composition and Diversity
Its important to not only be able to visualise the data but also understand relationships between barcodes/tags at the data level

### Diversity analysis
We can examine diversity in a few different ways. 
- Shannon
- Simpson / Inverse Simpson
- Gini
Each will be most applicable in different circumstances, however the Shannon diversity index is widely used to compare global diversity amongst populations of barcoded cells. 
```{r}
diversity <- calcDivIndexes(dge.filtered.collapsed$counts)
diversity
qplot(diversity$name, diversity$shannon) + theme_bw() + coord_flip()
```

### Comparing abundance
We can statistically test for barcodes / tags that are over / underrepresented in a group of samples relative to another using the internal edgeR framework. bartools contains a convenience wrapper for this functionality
```{r}
diff.bc <- compareAbundance(dge.filtered.collapsed, 
                 condition = dge.filtered.collapsed$samples$Treatment, 
                 condition_names = c("Vehicle", "High_dose"))
diff.bc
```

### Linegraphs of barcode abundance - NOT WORKING
```{r, eval = FALSE}
plotAbundanceLines(dge.filtered.collapsed, condition = dge.filtered.collapsed$samples$group, condition_names = c("Vehicle","High_dose"), plot_type = 'counts')
```


