<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bartools">
<title>bartools: scRNA-seq analysis guide • bartools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="bartools: scRNA-seq analysis guide">
<meta property="og:description" content="bartools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bartools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bartools_quickstart.html">bartools: quickstart guide</a>
    <a class="dropdown-item" href="../articles/bartools_single_cell_analysis.html">bartools: scRNA-seq analysis guide</a>
    <a class="dropdown-item" href="../articles/import_qc_single_cell.html">bartools: import and QC single-cell barcode data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>bartools: scRNA-seq analysis guide</h1>
                        <h4 data-toc-skip class="author">Henrietta
Holze</h4>
            
            <h4 data-toc-skip class="date">September 05, 2023</h4>
      
      
      <div class="d-none name"><code>bartools_single_cell_analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="bartools-single-cell-guide">Bartools single-cell guide<a class="anchor" aria-label="anchor" href="#bartools-single-cell-guide"></a>
</h2>
<p>The <code>bartools</code> package contains methods to simplify clone
level analyses from single-cell cellular barcoding datasets. The purpose
of this vignette is to highlight these capabilities of
<code>bartools</code>.</p>
<div class="section level3">
<h3 id="setup">0. Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>This vignette makes use of a simple test single cell experiment
object (SCE) included in the <code>bartools</code> package. The dataset
contains 100 cells each with 100 randomly sampled genes. Each cell has
lineage barcode information annotated in the
<code>test.sce$barcode</code> field.</p>
<p>Dimensionality reduction and clustering were previously performed
using <code>Seurat</code>. Cluster assignments are in the
<code>test.sce$seurat_clusters</code> field.</p>
<div class="section level4">
<h4 id="load-the-bartools-package">Load the <code>bartools</code> package<a class="anchor" aria-label="anchor" href="#load-the-bartools-package"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bartools</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: edgeR</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: limma</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>dev<span class="op">=</span><span class="st">"png"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="load-the-test-sce-dataset">Load the <code>test.sce</code> dataset<a class="anchor" aria-label="anchor" href="#load-the-test-sce-dataset"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span></span>
<span><span class="va">test.sce</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SingleCellExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: SummarizedExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:limma':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SingleCellExperiment'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:edgeR':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     cpm</span></span></code></pre>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 100 100 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(3): counts logcounts scaledata</span></span>
<span><span class="co">## rownames(100): Mrpl15 Lypla1 ... Tmem131 Cnga3</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(100): GCTACCTAGAGGCCAT ACTGCAAGTGATTCAC ... CGAGTTACAGCGACAA</span></span>
<span><span class="co">##   TTTATGCGTACAGTTC</span></span>
<span><span class="co">## colData names(27): orig.ident nCount_RNA ... seurat_clusters ident</span></span>
<span><span class="co">## reducedDimNames(3): PCA TSNE UMAP</span></span>
<span><span class="co">## mainExpName: RNA</span></span>
<span><span class="co">## altExpNames(1): HTO</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="analyse-dataset-metrics-using-plotmetrics">1. Analyse dataset metrics using plotMetrics<a class="anchor" aria-label="anchor" href="#analyse-dataset-metrics-using-plotmetrics"></a>
</h3>
<p>Single cell RNA sequencing datasets can reveal transcriptional
differences between groups of cells / celltypes that are of biological
importance. The <code>plotMetrics</code> function takes a single cell
object in Seurat or SingleCellExperiment format and plots a desired
continuous variable (e.g. number of transcripts or genes detected per
cell) split across any groups of interest. These groups could be
clusters, celltypes or, with lineage barcode information, individual
clones. Thus, <code>plotMetrics</code> allows us to go one step further
and examine biological differences between individual groups of
cells.</p>
<p>Here, the <code>group</code> parameter defines a grouping variable,
present as a column of metadata in the single cell object. the
<code>factor</code> parameter defines a continuous variable to plot per
level of the <code>group</code> parameter. the <code>threshold</code>
parameter defines the minimum number of cells required for each level of
the <code>group</code> parameter.</p>
<div class="section level4">
<h4 id="plotmetrics---clusters">plotMetrics - clusters<a class="anchor" aria-label="anchor" href="#plotmetrics---clusters"></a>
</h4>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMetrics.html">plotMetrics</a></span><span class="op">(</span><span class="va">test.sce</span>, group <span class="op">=</span> <span class="st">"seurat_clusters"</span>, factor <span class="op">=</span> <span class="st">"nCount_RNA"</span>, threshold <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-2-1.png" width="700">
Using the <code>trans</code> parameter, the x axis in
<code>plotMetrics</code> can be transformed using standard methods
available within <code>R</code></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMetrics.html">plotMetrics</a></span><span class="op">(</span><span class="va">test.sce</span>, group <span class="op">=</span> <span class="st">"seurat_clusters"</span>, factor <span class="op">=</span> <span class="st">"nCount_RNA"</span>, threshold <span class="op">=</span> <span class="fl">10</span>, trans <span class="op">=</span> <span class="st">"log10"</span><span class="op">)</span></span></code></pre></div>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="plotmetrics---barcodes">plotMetrics - barcodes<a class="anchor" aria-label="anchor" href="#plotmetrics---barcodes"></a>
</h4>
<p><code>plotMetrics</code> accepts any grouping variable available in
the sample metadata. Here we examine transcriptional differences between
clones using lineage barcode information.</p>
<p>NB: In this test dataset the number of cells per clone is small.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMetrics.html">plotMetrics</a></span><span class="op">(</span><span class="va">test.sce</span>, group <span class="op">=</span> <span class="st">"barcode"</span>, factor <span class="op">=</span> <span class="st">"nCount_RNA"</span>, threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="examine-number-of-cells-per-grouping-variable---plotcellspergroup">2. Examine number of cells per grouping variable -
plotCellsPerGroup<a class="anchor" aria-label="anchor" href="#examine-number-of-cells-per-grouping-variable---plotcellspergroup"></a>
</h3>
<p>We may also be interested in basic metrics like how many cells
comprise a grouping variable of interest such as cluster, celltype or
lineage barcode. The <code>plotCellsPerGroup</code> function allows this
to be easily plotted.</p>
<p>Groups above a user defined threshold are highlighted.</p>
<p>Here, the <code>group</code> parameter defines a grouping variable,
present as a column of metadata in the single cell object. the
<code>threshold</code> parameter defines the minimum number of cells.
Levels of the <code>group</code> parameter above this threshold will be
labelled.</p>
<div class="section level4">
<h4 id="plot-cells-per-cluster">plot cells per cluster<a class="anchor" aria-label="anchor" href="#plot-cells-per-cluster"></a>
</h4>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellsPerGroup.html">plotCellsPerGroup</a></span><span class="op">(</span><span class="va">test.sce</span>, group <span class="op">=</span> <span class="st">"seurat_clusters"</span>, threshold <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="plot-cells-per-lineage-barcode">plot cells per lineage barcode<a class="anchor" aria-label="anchor" href="#plot-cells-per-lineage-barcode"></a>
</h4>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellsPerGroup.html">plotCellsPerGroup</a></span><span class="op">(</span><span class="va">test.sce</span>, group <span class="op">=</span> <span class="st">"barcode"</span>, threshold <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="plot-distribution-of-cells-across-clusters---plotcellsinclusters">3. Plot distribution of cells across clusters -
plotCellsInClusters<a class="anchor" aria-label="anchor" href="#plot-distribution-of-cells-across-clusters---plotcellsinclusters"></a>
</h3>
<p>We may also be interested in the distribution of cells within certain
groups across levels of another group. For example we may want to reveal
the proportion of cells in a certain cell cycle phase across louvain
clusters within a single cell dataset. The
<code>plotCellsInClusters</code> function allows users to examine these
questions.</p>
<p>Here, the <code>group</code> parameter defines a grouping variable,
present as a column of metadata in the single cell object. the
<code>factor</code> parameter defines a level of <code>group</code> to
calculate percentage abundance within each level of <code>idents</code>.
the <code>idents</code> parameter defines a second grouping variable,
present as a column of metadata in the single cell object.</p>
<div class="section level4">
<h4 id="cell-cycle-phase-across-clusters">Cell cycle phase across clusters<a class="anchor" aria-label="anchor" href="#cell-cycle-phase-across-clusters"></a>
</h4>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellsInClusters.html">plotCellsInClusters</a></span><span class="op">(</span><span class="va">test.sce</span>, group <span class="op">=</span> <span class="st">"Phase"</span>, factor <span class="op">=</span> <span class="st">"G1"</span>, clusters <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 7 × 2</span></span></span>
<span><span class="co">##   seurat_clusters     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 0                  17</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 1                  15</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 2                  15</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> 3                   6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> 4                   4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> 5                   3</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> 7                   1</span></span></code></pre>
<pre><code><span><span class="co">## Warning in ggplot2::geom_histogram(stat = "identity"): Ignoring unknown</span></span>
<span><span class="co">## parameters: `binwidth`, `bins`, and `pad`</span></span></code></pre>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>The <code>plotCellsInClusters</code> framework can extend to any
discrete variables present in the dataset. Here we examine the
representation of lineage barcode BC_12904 across clusters</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellsInClusters.html">plotCellsInClusters</a></span><span class="op">(</span><span class="va">test.sce</span>, group <span class="op">=</span> <span class="st">"barcode"</span>, factor <span class="op">=</span> <span class="st">"BC_12904"</span>, clusters <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 2</span></span></span>
<span><span class="co">##   seurat_clusters     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 0                   2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 1                   1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 7                   1</span></span></code></pre>
<pre><code><span><span class="co">## Warning in ggplot2::geom_histogram(stat = "identity"): Ignoring unknown</span></span>
<span><span class="co">## parameters: `binwidth`, `bins`, and `pad`</span></span></code></pre>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="determine-enrichment-within-clusters---plotclusterenrichment">4. Determine enrichment within clusters - plotClusterEnrichment<a class="anchor" aria-label="anchor" href="#determine-enrichment-within-clusters---plotclusterenrichment"></a>
</h3>
<p>To check whether a clone is enriched in a cluster or cell type, we
can perform a hypergeometric test using the
<code>plotClusterEnrichment</code> function. Here we test for enrichment
of cells in G2M cell cycle phase across louvain clusters.</p>
<p>Here, the <code>group</code> parameter defines a grouping variable,
present as a column of metadata in the single cell object. the
<code>factor</code> parameter defines a level of the <code>group</code>
parameter to test for enrichment within each level of
<code>idents</code>. the <code>idents</code> parameter defines a second
grouping variable to test for enrichment of <code>factor</code> at each
level. the <code>threshold</code> parameter defines a p-value threshold
for the hypergeometric test.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotClusterEnrichment.html">plotClusterEnrichment</a></span><span class="op">(</span></span>
<span>  <span class="va">test.sce</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"Phase"</span>,</span>
<span>  factor <span class="op">=</span> <span class="st">"G2M"</span>,</span>
<span>  clusters <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  order <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The following ident levels had no observations and were removed: 6</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_0</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all G2M cells: 2</span></span></code></pre>
<pre><code><span><span class="co">## universe: 98</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 24</span></span></code></pre>
<pre><code><span><span class="co">## cluster G2M cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.424242424242424"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_1</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all G2M cells: 2</span></span></code></pre>
<pre><code><span><span class="co">## universe: 98</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 27</span></span></code></pre>
<pre><code><span><span class="co">## cluster G2M cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.469090909090909"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_2</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all G2M cells: 2</span></span></code></pre>
<pre><code><span><span class="co">## universe: 98</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 28</span></span></code></pre>
<pre><code><span><span class="co">## cluster G2M cells: 1</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.0763636363636364"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_3</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all G2M cells: 2</span></span></code></pre>
<pre><code><span><span class="co">## universe: 98</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 10</span></span></code></pre>
<pre><code><span><span class="co">## cluster G2M cells: 1</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.0090909090909091"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_4</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all G2M cells: 2</span></span></code></pre>
<pre><code><span><span class="co">## universe: 98</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 5</span></span></code></pre>
<pre><code><span><span class="co">## cluster G2M cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.0979797979797982"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_5</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all G2M cells: 2</span></span></code></pre>
<pre><code><span><span class="co">## universe: 98</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 5</span></span></code></pre>
<pre><code><span><span class="co">## cluster G2M cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.0979797979797982"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_7</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all G2M cells: 2</span></span></code></pre>
<pre><code><span><span class="co">## universe: 98</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 1</span></span></code></pre>
<pre><code><span><span class="co">## cluster G2M cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.0200000000000001"</span></span></code></pre>
<p><img src="bartools_single_cell_analysis_files/figure-html/unnamed-chunk-9-1.png" width="700">
## 5. Session Info</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS 14.1.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SingleCellExperiment_1.20.1 SummarizedExperiment_1.28.0</span></span>
<span><span class="co">##  [3] Biobase_2.58.0              GenomicRanges_1.50.2       </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.34.9         IRanges_2.32.0             </span></span>
<span><span class="co">##  [7] S4Vectors_0.36.2            BiocGenerics_0.44.0        </span></span>
<span><span class="co">##  [9] MatrixGenerics_1.10.0       matrixStats_1.0.0          </span></span>
<span><span class="co">## [11] bartools_1.0.0              ggplot2_3.4.4              </span></span>
<span><span class="co">## [13] edgeR_3.40.2                limma_3.54.2               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] viridis_0.6.4          sass_0.4.7             viridisLite_0.4.2     </span></span>
<span><span class="co">##  [4] jsonlite_1.8.7         splines_4.2.2          bslib_0.6.0           </span></span>
<span><span class="co">##  [7] highr_0.10             GenomeInfoDbData_1.2.9 ggrepel_0.9.4         </span></span>
<span><span class="co">## [10] yaml_2.3.7             pillar_1.9.0           lattice_0.21-9        </span></span>
<span><span class="co">## [13] glue_1.6.2             digest_0.6.33          XVector_0.38.0        </span></span>
<span><span class="co">## [16] colorspace_2.1-0       htmltools_0.5.7        Matrix_1.6-1.1        </span></span>
<span><span class="co">## [19] pkgconfig_2.0.3        zlibbioc_1.44.0        purrr_1.0.2           </span></span>
<span><span class="co">## [22] scales_1.2.1           tibble_3.2.1           mgcv_1.9-0            </span></span>
<span><span class="co">## [25] farver_2.1.1           generics_0.1.3         cachem_1.0.8          </span></span>
<span><span class="co">## [28] withr_2.5.2            cli_3.6.1              magrittr_2.0.3        </span></span>
<span><span class="co">## [31] memoise_2.0.1          evaluate_0.23          fs_1.6.3              </span></span>
<span><span class="co">## [34] fansi_1.0.5            nlme_3.1-163           MASS_7.3-60           </span></span>
<span><span class="co">## [37] vegan_2.6-4            textshaping_0.3.6      tools_4.2.2           </span></span>
<span><span class="co">## [40] ineq_0.2-13            lifecycle_1.0.4        stringr_1.5.1         </span></span>
<span><span class="co">## [43] munsell_0.5.0          locfit_1.5-9.8         DelayedArray_0.24.0   </span></span>
<span><span class="co">## [46] cluster_2.1.4          compiler_4.2.2         pkgdown_2.0.7         </span></span>
<span><span class="co">## [49] jquerylib_0.1.4        systemfonts_1.0.5      rlang_1.1.2           </span></span>
<span><span class="co">## [52] grid_4.2.2             RCurl_1.98-1.12        rstudioapi_0.15.0     </span></span>
<span><span class="co">## [55] labeling_0.4.3         bitops_1.0-7           rmarkdown_2.25        </span></span>
<span><span class="co">## [58] gtable_0.3.4           R6_2.5.1               gridExtra_2.3         </span></span>
<span><span class="co">## [61] knitr_1.45             dplyr_1.1.4            fastmap_1.1.1         </span></span>
<span><span class="co">## [64] utf8_1.2.4             rprojroot_2.0.4        ragg_1.2.5            </span></span>
<span><span class="co">## [67] permute_0.9-7          desc_1.4.2             stringi_1.8.1         </span></span>
<span><span class="co">## [70] parallel_4.2.2         Rcpp_1.0.11            vctrs_0.6.4           </span></span>
<span><span class="co">## [73] tidyselect_1.2.0       xfun_0.41</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Dane Vassiliadis, Henrietta Holze.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
