<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="bartools">
<title>bartools: import and QC single-cell barcode data • bartools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="bartools: import and QC single-cell barcode data">
<meta property="og:description" content="bartools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">bartools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bartools_quickstart.html">bartools: quickstart guide</a>
    <a class="dropdown-item" href="../articles/bartools_single_cell_analysis.html">bartools: scRNA-seq analysis guide</a>
    <a class="dropdown-item" href="../articles/import_qc_single_cell.html">bartools: import and QC single-cell barcode data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>bartools: import and QC single-cell barcode data</h1>
                        <h4 data-toc-skip class="author">Henrietta
Holze</h4>
            
            <h4 data-toc-skip class="date">September 05, 2023</h4>
      
      
      <div class="d-none name"><code>import_qc_single_cell.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="import-and-qc-of-single-cell-barcode-data">Import and QC of single-cell barcode data<a class="anchor" aria-label="anchor" href="#import-and-qc-of-single-cell-barcode-data"></a>
</h2>
<p>This vignette shows how to 1. Import the barcode counts table
produced by BARtab single-cell workflow 2. Creat QC plots and filter
barcodes 3. Add barcode counts to a Seurat object 4. Analyze and
visualize barcoded single-cell RNA-seq data</p>
<p>If filtering has already been performed by BARtab, using
<code>umi_count_filter</code> and <code>umi_fraction_filter</code>
parameters, skip to section 2.<br>
If the QC thresholds need to be revised, follow section 1.</p>
<div class="section level3">
<h3 id="load-the-bartools-package">0. Load the <code>bartools</code> package<a class="anchor" aria-label="anchor" href="#load-the-bartools-package"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bartools</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: edgeR</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: limma</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>dev<span class="op">=</span><span class="st">"png"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="import-dna-barcode-count-data">1. Import DNA barcode count data<a class="anchor" aria-label="anchor" href="#import-dna-barcode-count-data"></a>
</h3>
<p>Raw barcode count data can be thought of similarly to raw
integer-based count data from other count based experiments such as
RNA-sequencing. For these data types the <code>edgeR</code> package
provides an efficient <code>DGEList</code> object structure to store
sample counts and associated metadata. <code>bartools</code> makes use
of this object structure to store and process DNA barcode counts.</p>
<div class="section level4">
<h4 id="an-example-barcoding-experiment">An example barcoding experiment<a class="anchor" aria-label="anchor" href="#an-example-barcoding-experiment"></a>
</h4>
<p>For this section we will make use of a hypothetical DNA barcoding
dataset based on recent unpublished data from the Dawson lab
investigating the response of acute myeloid leukaemia (AML) cells to a
novel class of MYST acetyltransferase inhibitor described recently in <a href="https://dx.doi.org/10.1038/s41586-019-1835-6" class="external-link">MacPherson et
al. Nature 2019</a>.</p>
<p>AML cells were cultured <em>in vitro</em>, barcoded using a
lentiviral based barcoding library called <a href="https://dx.doi.org/10.1038/s41586-021-04206-7" class="external-link">SPLINTR</a>, and
transplanted into three groups of C57BL/6J mice with daily dosing of
MYST inhibitor at low or high dose or a corresponding vehicle
control.</p>
<p>Barcode containing cells were harvested from the bone marrow of
diseased mice and sequenced in technical replicate.</p>
<p>To follow along with this vignette the raw counts tables and sample
metadata are included in the <code>bartools</code> package.</p>
</div>
</div>
<div class="section level3">
<h3 id="qc-and-filter-barcodes">2. QC and filter barcodes<a class="anchor" aria-label="anchor" href="#qc-and-filter-barcodes"></a>
</h3>
<p>First, we load the results of BARtab. Below is an example of how we
can load this data using the <code>readBartabCounts</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_path</span> <span class="op">&lt;-</span> <span class="st">"full/path/to/BARtab/umi.counts.tsv"</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readBartabCounts.html">readBartabCounts</a></span><span class="op">(</span><span class="va">counts_path</span><span class="op">)</span></span></code></pre></div>
<p>For now, users can follow along with this vignette using the test
dataset included in <code>bartools</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">test.bartab.sc</span><span class="op">)</span></span></code></pre></div>
<p>This unfiltered barcode counts table is in long format and
potentially has multiple rows per cell ID if multiple barcodes per cell
are detected.</p>
<p>We assess the quality of the data by looking at the number of
barcodes detected per cell and the number of UMIs supporting the most
frequent barcode per cell.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBarcodesPerCell.html">plotBarcodesPerCell</a></span><span class="op">(</span><span class="va">test.bartab.sc</span><span class="op">)</span></span></code></pre></div>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotUmiPerBarcode.html">plotUmiPerBarcode</a></span><span class="op">(</span><span class="va">test.bartab.sc</span><span class="op">)</span></span></code></pre></div>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>In order to identify a suitable threshold for the minimum number of
UMIs supporting a barcode, we can plot the cummulative sum of how many
cells would pass each threshold.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotUmiFilterThresholds.html">plotUmiFilterThresholds</a></span><span class="op">(</span><span class="va">test.bartab.sc</span><span class="op">)</span></span></code></pre></div>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>If we would remove barcodes that are supported by a single UMI, we
would lose half the cells. Therefore, we do not filter based on minimum
number of UMIs.</p>
<p>However, we can remove minor barcodes from cells, i.e. barcodes that
have less than half the number of supporting UMIs than the major barcode
in the cell.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterBarcodes.html">filterBarcodes</a></span><span class="op">(</span><span class="va">test.bartab.sc</span>, umiCountFilter <span class="op">=</span> <span class="fl">2</span>, umiFractionFilter <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p>The following plot shows that the number of cells with multiple
barcodes detected is now reduced.</p>
<p>If we would want to keep only the most frequent barcode per cell, we
could do so with <code>umi_fraction_filter = 1</code>. Ties will be
kept.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBarcodesPerCell.html">plotBarcodesPerCell</a></span><span class="op">(</span><span class="va">counts_filtered</span><span class="op">)</span></span></code></pre></div>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Finally, we aggregate barcodes per cell by concatenating barcodes and
UMIs with <code>;</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_agg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateBarcodes.html">aggregateBarcodes</a></span><span class="op">(</span><span class="va">counts_filtered</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">counts_agg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             cellid               barcode bc.umi.count</span></span>
<span><span class="co">## 1 AAACCCAAGTTTGTCG mCHERRY_Barcode_12409            4</span></span>
<span><span class="co">## 2 AAACCCACAAGAGCTG mCHERRY_Barcode_14566            3</span></span>
<span><span class="co">## 3 AAACCCATCAGACCCG mCHERRY_Barcode_12904            3</span></span>
<span><span class="co">## 4 AAACGCTCATATAGCC  mCHERRY_Barcode_1614            3</span></span>
<span><span class="co">## 5 AAACGCTTCTCAGGCG  mCHERRY_Barcode_1614            2</span></span>
<span><span class="co">## 6 AAACGCTTCTGACAGT mCHERRY_Barcode_11745            3</span></span></code></pre>
<p>This allows to add the barcode data into the metadata of a
<code>Seurat</code> or <code>SingleCellExperiment</code> object.</p>
</div>
<div class="section level3">
<h3 id="add-barcode-data-to-seurat-or-singlecellexperiment-object">2. Add barcode data to Seurat or SingleCellExperiment object<a class="anchor" aria-label="anchor" href="#add-barcode-data-to-seurat-or-singlecellexperiment-object"></a>
</h3>
<div class="section level4">
<h4 id="barcode-metadata-import-into-seurat">Barcode metadata import into Seurat<a class="anchor" aria-label="anchor" href="#barcode-metadata-import-into-seurat"></a>
</h4>
<p>Generate Seurat object from Cell Ranger scRNA expression matrix.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">expression.data</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"../../bartools_sc_example/GEX_filtered_feature_bc_matrix/"</span>, strip.suffix <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">expression.data</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min.features <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  project <span class="op">=</span> <span class="st">"bartools_example"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sc</span></span></code></pre></div>
<p>Add SPLINTR barcode annotation to metadata. Make sure the cell IDs
match. Potentially suffix needs to be added to the BARtab results
(e.g. <code>-1</code>). Add barcode and UMI count columns to the
metadata of the SCE object.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_agg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">counts_agg</span><span class="op">$</span><span class="va">cellid</span></span>
<span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AddMetaData.html" class="external-link">AddMetaData</a></span><span class="op">(</span><span class="va">sc</span>, <span class="va">counts_agg</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="barcode-metadata-import-into-singlecellexperiment">Barcode metadata import into SingleCellExperiment<a class="anchor" aria-label="anchor" href="#barcode-metadata-import-into-singlecellexperiment"></a>
</h4>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcode"</span>, <span class="st">"bc.umi.count"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">counts_agg</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcode"</span>, <span class="st">"bc.umi.count"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SingleCellExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: SummarizedExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:limma':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SingleCellExperiment'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:edgeR':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     cpm</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 27 columns</span></span>
<span><span class="co">##                        orig.ident nCount_RNA nFeature_RNA percent.mito</span></span>
<span><span class="co">##                          &lt;factor&gt;  &lt;numeric&gt;    &lt;integer&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## GCTACCTAGAGGCCAT bartools_example      14721         3122      6.86095</span></span>
<span><span class="co">## ACTGCAAGTGATTCAC bartools_example      20112         3497      4.52963</span></span>
<span><span class="co">## ATTTCTGGTTAGAGTA bartools_example      23773         4118      7.08367</span></span>
<span><span class="co">## GAAGTAAAGTCATGGG bartools_example       6155         1942      4.17547</span></span>
<span><span class="co">## CGGGACTCAAACCGGA bartools_example      30609         4583      4.80904</span></span>
<span><span class="co">## TGCATCCCACCTGCTT bartools_example       9633         2401      4.93097</span></span>
<span><span class="co">##                  nCount_HTO nFeature_HTO   HTO_maxID HTO_secondID HTO_margin</span></span>
<span><span class="co">##                   &lt;numeric&gt;    &lt;integer&gt; &lt;character&gt;  &lt;character&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## GCTACCTAGAGGCCAT       1098            2       MA9-1        MA9-2   1.426341</span></span>
<span><span class="co">## ACTGCAAGTGATTCAC        748            2       MA9-1        MA9-2   1.483240</span></span>
<span><span class="co">## ATTTCTGGTTAGAGTA        788            2       MA9-1        MA9-2   1.468272</span></span>
<span><span class="co">## GAAGTAAAGTCATGGG        351            2       MA9-1        MA9-2   0.986854</span></span>
<span><span class="co">## CGGGACTCAAACCGGA        731            2       MA9-1        MA9-2   1.398757</span></span>
<span><span class="co">## TGCATCCCACCTGCTT        790            2       MA9-1        MA9-2   1.463313</span></span>
<span><span class="co">##                  HTO_classification HTO_classification.global  hash.ID</span></span>
<span><span class="co">##                         &lt;character&gt;               &lt;character&gt; &lt;factor&gt;</span></span>
<span><span class="co">## GCTACCTAGAGGCCAT              MA9-1                   Singlet    MA9-1</span></span>
<span><span class="co">## ACTGCAAGTGATTCAC              MA9-1                   Singlet    MA9-1</span></span>
<span><span class="co">## ATTTCTGGTTAGAGTA              MA9-1                   Singlet    MA9-1</span></span>
<span><span class="co">## GAAGTAAAGTCATGGG              MA9-1                   Singlet    MA9-1</span></span>
<span><span class="co">## CGGGACTCAAACCGGA              MA9-1                   Singlet    MA9-1</span></span>
<span><span class="co">## TGCATCCCACCTGCTT              MA9-1                   Singlet    MA9-1</span></span>
<span><span class="co">##                      barcode bc.umi.count     detected doubletBarcode</span></span>
<span><span class="co">##                  &lt;character&gt;  &lt;character&gt;  &lt;character&gt;    &lt;character&gt;</span></span>
<span><span class="co">## GCTACCTAGAGGCCAT          NA           NA     detected        singlet</span></span>
<span><span class="co">## ACTGCAAGTGATTCAC          NA           NA     detected        singlet</span></span>
<span><span class="co">## ATTTCTGGTTAGAGTA          NA           NA     detected        singlet</span></span>
<span><span class="co">## GAAGTAAAGTCATGGG          NA           NA not.detected        unknown</span></span>
<span><span class="co">## CGGGACTCAAACCGGA          NA           NA     detected        singlet</span></span>
<span><span class="co">## TGCATCCCACCTGCTT          NA           NA     detected        singlet</span></span>
<span><span class="co">##                  MA9_1_HTO_raw MA9_2_HTO_raw MA9_1_HTO_norm MA9_2_HTO_norm</span></span>
<span><span class="co">##                      &lt;numeric&gt;     &lt;numeric&gt;      &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">## GCTACCTAGAGGCCAT          1017            81        1.88644      0.4600971</span></span>
<span><span class="co">## ACTGCAAGTGATTCAC           729            19        1.61167      0.1284293</span></span>
<span><span class="co">## ATTTCTGGTTAGAGTA           761            27        1.64620      0.1779307</span></span>
<span><span class="co">## GAAGTAAAGTCATGGG           341            10        1.05650      0.0696444</span></span>
<span><span class="co">## CGGGACTCAAACCGGA           703            28        1.58271      0.1839496</span></span>
<span><span class="co">## TGCATCCCACCTGCTT           762            28        1.64726      0.1839496</span></span>
<span><span class="co">##                     S.Score G2M.Score       Phase old.ident RNA_snn_res.0.3</span></span>
<span><span class="co">##                   &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;  &lt;factor&gt;        &lt;factor&gt;</span></span>
<span><span class="co">## GCTACCTAGAGGCCAT  0.5712484  -2.98143           S   Singlet               0</span></span>
<span><span class="co">## ACTGCAAGTGATTCAC -0.7336216  -2.62838          G1   Singlet               3</span></span>
<span><span class="co">## ATTTCTGGTTAGAGTA  0.7472527  -3.80758           S   Singlet               1</span></span>
<span><span class="co">## GAAGTAAAGTCATGGG -0.0413439  -1.08835          G1   Singlet               1</span></span>
<span><span class="co">## CGGGACTCAAACCGGA -0.6348105  -3.49328          G1   Singlet               2</span></span>
<span><span class="co">## TGCATCCCACCTGCTT -0.2304990  -1.44251          G1   Singlet               2</span></span>
<span><span class="co">##                  seurat_clusters    ident</span></span>
<span><span class="co">##                         &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">## GCTACCTAGAGGCCAT               0        0</span></span>
<span><span class="co">## ACTGCAAGTGATTCAC               3        3</span></span>
<span><span class="co">## ATTTCTGGTTAGAGTA               1        1</span></span>
<span><span class="co">## GAAGTAAAGTCATGGG               1        1</span></span>
<span><span class="co">## CGGGACTCAAACCGGA               2        2</span></span>
<span><span class="co">## TGCATCCCACCTGCTT               2        2</span></span></code></pre>
<p>Group barcode metadata into detected and not.detected</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test.sce</span><span class="op">$</span><span class="va">detected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>, <span class="st">"not.detected"</span>, <span class="st">"detected"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">$</span><span class="va">detected</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## not.detected </span></span>
<span><span class="co">##          100</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="analyze-and-visualize-barcoded-scrna-seq-data">4. Analyze and visualize barcoded scRNA-seq data<a class="anchor" aria-label="anchor" href="#analyze-and-visualize-barcoded-scrna-seq-data"></a>
</h3>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Percentage of cells with no barcode detected"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Percentage of cells with no barcode detected</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 29</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Percentage of cells with a barcode detected"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Percentage of cells with a barcode detected</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">test.sce</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 71</span></span></code></pre>
<p>We can repeat the QC plot to check how many barcodes were detected in
a cell, including no barcodes detected.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBarcodesPerCell.html">plotBarcodesPerCell</a></span><span class="op">(</span><span class="va">test.sce</span>, aggregated <span class="op">=</span> <span class="cn">T</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span></code></pre></div>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>We can visualize the distribution of number of cells per clone.</p>
<p>For that, plot all clones, ordered by their size. A clone is here
defined as barcode or unique combination of barcodes.</p>
<p>Usually, more than one barcode detected in a cell could mean a
doublet. The number of cells for those clones tend to be low. Seeing
more cells per clone with more than one barcode suggests it is truly
from two integration events.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellsPerGroup.html">plotCellsPerGroup</a></span><span class="op">(</span></span>
<span>  <span class="va">test.sce</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"barcode"</span>,</span>
<span>  order <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  label <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">","</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>To check whether a clone is enriched in a cluster or cell type, we
can perform a hypergeometric test.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotClusterEnrichment.html">plotClusterEnrichment</a></span><span class="op">(</span></span>
<span>  <span class="va">test.sce</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"barcode"</span>,</span>
<span>  factor <span class="op">=</span> <span class="st">"BC_1614"</span>,</span>
<span>  clusters <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  order <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The following ident levels had no observations and were removed: 6</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_0</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## universe: 97</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 24</span></span></code></pre>
<pre><code><span><span class="co">## cluster BC_1614 cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.565244279529994"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_1</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## universe: 97</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 27</span></span></code></pre>
<pre><code><span><span class="co">## cluster BC_1614 cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.615361781076067"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_2</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## universe: 97</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 28</span></span></code></pre>
<pre><code><span><span class="co">## cluster BC_1614 cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.631168831168831"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_3</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## universe: 97</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 10</span></span></code></pre>
<pre><code><span><span class="co">## cluster BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_4</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## universe: 97</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 5</span></span></code></pre>
<pre><code><span><span class="co">## cluster BC_1614 cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.14400123685838"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_5</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## universe: 97</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 5</span></span></code></pre>
<pre><code><span><span class="co">## cluster BC_1614 cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.14400123685838"</span></span></code></pre>
<pre><code><span><span class="co">## ---</span></span></code></pre>
<pre><code><span><span class="co">## cluster_7</span></span></code></pre>
<pre><code><span><span class="co">## all cells: 100</span></span></code></pre>
<pre><code><span><span class="co">## all BC_1614 cells: 3</span></span></code></pre>
<pre><code><span><span class="co">## universe: 97</span></span></code></pre>
<pre><code><span><span class="co">## cluster total cells: 1</span></span></code></pre>
<pre><code><span><span class="co">## cluster BC_1614 cells: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Hypergeometric test p-value: 0.03"</span></span></code></pre>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellsInClusters.html">plotCellsInClusters</a></span><span class="op">(</span><span class="va">test.sce</span>, </span>
<span>                    group <span class="op">=</span> <span class="st">"barcode"</span>, </span>
<span>                    factor <span class="op">=</span> <span class="st">"BC_12904"</span>, </span>
<span>                    clusters <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 2</span></span></span>
<span><span class="co">##   seurat_clusters     n</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 0                   2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 1                   1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 7                   1</span></span></code></pre>
<pre><code><span><span class="co">## Warning in ggplot2::geom_histogram(stat = "identity"): Ignoring unknown</span></span>
<span><span class="co">## parameters: `binwidth`, `bins`, and `pad`</span></span></code></pre>
<p><img src="import_qc_single_cell_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="session-info">5. Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS 14.1.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SingleCellExperiment_1.20.1 SummarizedExperiment_1.28.0</span></span>
<span><span class="co">##  [3] Biobase_2.58.0              GenomicRanges_1.50.2       </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.34.9         IRanges_2.32.0             </span></span>
<span><span class="co">##  [7] S4Vectors_0.36.2            BiocGenerics_0.44.0        </span></span>
<span><span class="co">##  [9] MatrixGenerics_1.10.0       matrixStats_1.0.0          </span></span>
<span><span class="co">## [11] bartools_1.0.0              ggplot2_3.4.4              </span></span>
<span><span class="co">## [13] edgeR_3.40.2                limma_3.54.2               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] viridis_0.6.4          sass_0.4.7             viridisLite_0.4.2     </span></span>
<span><span class="co">##  [4] jsonlite_1.8.7         splines_4.2.2          bslib_0.6.0           </span></span>
<span><span class="co">##  [7] highr_0.10             GenomeInfoDbData_1.2.9 ggrepel_0.9.4         </span></span>
<span><span class="co">## [10] yaml_2.3.7             pillar_1.9.0           lattice_0.21-9        </span></span>
<span><span class="co">## [13] glue_1.6.2             digest_0.6.33          XVector_0.38.0        </span></span>
<span><span class="co">## [16] colorspace_2.1-0       htmltools_0.5.7        Matrix_1.6-1.1        </span></span>
<span><span class="co">## [19] pkgconfig_2.0.3        zlibbioc_1.44.0        purrr_1.0.2           </span></span>
<span><span class="co">## [22] scales_1.2.1           tibble_3.2.1           mgcv_1.9-0            </span></span>
<span><span class="co">## [25] generics_0.1.3         farver_2.1.1           cachem_1.0.8          </span></span>
<span><span class="co">## [28] withr_2.5.2            cli_3.6.1              magrittr_2.0.3        </span></span>
<span><span class="co">## [31] memoise_2.0.1          evaluate_0.23          fs_1.6.3              </span></span>
<span><span class="co">## [34] fansi_1.0.5            nlme_3.1-163           MASS_7.3-60           </span></span>
<span><span class="co">## [37] vegan_2.6-4            textshaping_0.3.6      tools_4.2.2           </span></span>
<span><span class="co">## [40] ineq_0.2-13            data.table_1.14.8      lifecycle_1.0.4       </span></span>
<span><span class="co">## [43] stringr_1.5.1          munsell_0.5.0          locfit_1.5-9.8        </span></span>
<span><span class="co">## [46] DelayedArray_0.24.0    cluster_2.1.4          compiler_4.2.2        </span></span>
<span><span class="co">## [49] pkgdown_2.0.7          jquerylib_0.1.4        systemfonts_1.0.5     </span></span>
<span><span class="co">## [52] rlang_1.1.2            grid_4.2.2             RCurl_1.98-1.12       </span></span>
<span><span class="co">## [55] rstudioapi_0.15.0      bitops_1.0-7           labeling_0.4.3        </span></span>
<span><span class="co">## [58] rmarkdown_2.25         gtable_0.3.4           R6_2.5.1              </span></span>
<span><span class="co">## [61] gridExtra_2.3          knitr_1.45             dplyr_1.1.4           </span></span>
<span><span class="co">## [64] fastmap_1.1.1          utf8_1.2.4             rprojroot_2.0.4       </span></span>
<span><span class="co">## [67] ragg_1.2.5             permute_0.9-7          desc_1.4.2            </span></span>
<span><span class="co">## [70] stringi_1.8.1          parallel_4.2.2         Rcpp_1.0.11           </span></span>
<span><span class="co">## [73] vctrs_0.6.4            tidyselect_1.2.0       xfun_0.41</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Dane Vassiliadis, Henrietta Holze.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
